qutebrowser/browser/greasemonkey.py

```diff
@@ -0,0 +1,289 @@
+
+    if(!details.url) {{
+        throw("GM_xmlhttpRequest requires an URL.");
+    }}
+
+    // build XMLHttpRequest object
+    var oXhr = new XMLHttpRequest;
+    // run it
+    if(oXhr) {{
+        if("onreadystatechange" in details)
+            oXhr.onreadystatechange = function() {{
+                details.onreadystatechange(oXhr)
+            }};
+        if("onload" in details)
+            oXhr.onload = function() {{ details.onload(oXhr) }};
+        if("onerror" in details)
+            oXhr.onerror = function() {{ details.onerror(oXhr) }};
+
+        oXhr.open(details.method, details.url, true);
+
+        if("headers" in details)
+            for(var header in details.headers)
+                oXhr.setRequestHeader(header, details.headers[header]);
+
+        if("data" in details)
+            oXhr.send(details.data);
+        else
+            oXhr.send();
+    }} else
+        throw ("This Browser is not supported, please upgrade.")
+}}
+
+function GM_addStyle(/* String */ styles) {{
+    var head = document.getElementsByTagName("head")[0];
+    if (head === undefined) {{
+        document.onreadystatechange = function() {{
+            if (document.readyState == "interactive") {{
+                var oStyle = document.createElement("style");
+                oStyle.setAttribute("type", "text\/css");
+                oStyle.appendChild(document.createTextNode(styles));
+                document.getElementsByTagName("head")[0].appendChild(oStyle);
+            }}
+        }}
+    }}
+    else {{
+        var oStyle = document.createElement("style");
+        oStyle.setAttribute("type", "text\/css");
+        oStyle.appendChild(document.createTextNode(styles));
+        head.appendChild(oStyle);
+    }}
+}}
+
+unsafeWindow = window;
+"""
+
+    def __init__(self, properties, code):
+        self._code = code
+        self._includes = []
+        self._excludes = []
+        self._description = None
+        self._name = None
+        self._run_at = None
+        for name, value in properties:
+            if name == 'name':
+                self._name = value
+            if name == 'description':
+                self._description = value
+            if name in ['include', 'match']:
+                self._includes.append(value)
+            if name in ['exclude', 'exclude_match']:
+                self._excludes.append(value)
+            if name == 'run-at':
+                self._run_at = value
+
+    HEADER_REGEX = r'\/\/ ==UserScript==.|\n+\/\/ ==\/UserScript==\n'
+    PROPS_REGEX = r'\/\/ @(?P<prop>[^\s]+)\s+(?P<val>.+)'
+
+    @classmethod
+    def parse(cls, source):
+        props, _code = re.split(cls.HEADER_REGEX, source)
+        return cls(re.findall(cls.PROPS_REGEX, props), source)
+
+    def includes(self):
+        return self._includes
+
+    def excludes(self):
+        return self._excludes
+
+    def name(self):
+        return self._name
+
+    def run_at(self):
+        return self._run_at
```

Comment:
I don't think those getter functions are useful - the properties should just not have a `_` and then be accessed like that from elsewhere.

Reply:
Yay, less docstrings!